File: mobile/packages/expo-mobile-web-shell/ios/MobileWebExactJson.swift
import Foundation

private let mobileWebJsonDepthLimit = 32

func isExactMobileWebJsonDocument(_ value: String) -> Bool {
  var parser = MobileWebExactJsonParser(bytes: Array(value.utf8))
  return parser.parse()
}

private struct MobileWebExactJsonParser {
  let bytes: [UInt8]
  var index = 0

  mutating func parse() -> Bool {
    skipWhitespace()
    guard parseValue(depth: 0) else { return false }
    skipWhitespace()
    return index == bytes.count
  }

  private mutating func parseValue(depth: Int) -> Bool {
    guard depth <= mobileWebJsonDepthLimit, index < bytes.count else { return false }
    switch bytes[index] {
    case 0x7B:
      return parseObject(depth: depth)
    case 0x5B:
      return parseArray(depth: depth)
    case 0x22:
      return parseStringLiteral() != nil
    case 0x74:
      return consume("true")
    case 0x66:
      return consume("false")
    case 0x6E:
      return consume("null")
    default:
      return parseNumber()
    }
  }

  private mutating func parseObject(depth: Int) -> Bool {
    guard consumeByte(0x7B) else { return false }
    skipWhitespace()
    if consumeByte(0x7D) { return true }
    var keys = Set<String>()
    while true {
      guard
        let literal = parseStringLiteral(),
        let key = decodeString(literal),
        keys.insert(key).inserted
      else {
        return false
      }
      skipWhitespace()
      guard consumeByte(0x3A) else { return false }
      skipWhitespace()
      guard parseValue(depth: depth + 1) else { return false }
      skipWhitespace()
      if consumeByte(0x7D) { return true }
      guard consumeByte(0x2C) else { return false }
      skipWhitespace()
    }
  }

  private mutating func parseArray(depth: Int) -> Bool {
    guard consumeByte(0x5B) else { return false }
    skipWhitespace()
    if consumeByte(0x5D) { return true }
    while true {
      guard parseValue(depth: depth + 1) else { return false }
      skipWhitespace()
      if consumeByte(0x5D) { return true }
      guard consumeByte(0x2C) else { return false }
      skipWhitespace()
    }
  }

  private mutating func parseStringLiteral() -> Range<Int>? {
    guard consumeByte(0x22) else { return nil }
    let start = index - 1
    while index < bytes.count {
      let byte = bytes[index]
      index += 1
      if byte == 0x22 { return start..<index }
      if byte < 0x20 { return nil }
      guard byte == 0x5C else { continue }
      guard index < bytes.count else { return nil }
      let escaped = bytes[index]
      index += 1
      if escaped == 0x75 {
        guard let codeUnit = parseUnicodeCodeUnit() else { return nil }
        if (0xD800...0xDBFF).contains(codeUnit) {
          guard
            consumeByte(0x5C),
            consumeByte(0x75),
            let lowSurrogate = parseUnicodeCodeUnit(),
            (0xDC00...0xDFFF).contains(lowSurrogate)
          else {
            return nil
          }
        } else if (0xDC00...0xDFFF).contains(codeUnit) {
          return nil
        }
      } else if ![0x22, 0x5C, 0x2F, 0x62, 0x66, 0x6E, 0x72, 0x74].contains(escaped) {
        return nil
      }
    }
    return nil
  }

  private mutating func parseUnicodeCodeUnit() -> UInt16? {
    guard index + 4 <= bytes.count else { return nil }
    var codeUnit: UInt16 = 0
    for byte in bytes[index..<(index + 4)] {
      guard let digit = hexValue(byte) else { return nil }
      codeUnit = (codeUnit << 4) | UInt16(digit)
    }
    index += 4
    return codeUnit
  }

  private mutating func parseNumber() -> Bool {
    let start = index
    _ = consumeByte(0x2D)
    guard index < bytes.count else { return false }
    if consumeByte(0x30) {
      if index < bytes.count, isDigit(bytes[index]) { return false }
    } else {
      guard consumeDigits(firstMustBeNonzero: true) else { return false }
    }
    if consumeByte(0x2E), !consumeDigits(firstMustBeNonzero: false) { return false }
    if index < bytes.count, bytes[index] == 0x65 || bytes[index] == 0x45 {
      index += 1
      if index < bytes.count, bytes[index] == 0x2B || bytes[index] == 0x2D {
        index += 1
      }
      guard consumeDigits(firstMustBeNonzero: false) else { return false }
    }
    return index > start
  }

  private mutating func consumeDigits(firstMustBeNonzero: Bool) -> Bool {
    guard
      index < bytes.count,
      firstMustBeNonzero ? isNonzeroDigit(bytes[index]) : isDigit(bytes[index])
    else {
      return false
    }
    index += 1
    while index < bytes.count, isDigit(bytes[index]) { index += 1 }
    return true
  }

  private mutating func consume(_ value: String) -> Bool {
    let expected = Array(value.utf8)
    guard bytes[index...].starts(with: expected) else { return false }
    index += expected.count
    return true
  }

  private mutating func consumeByte(_ byte: UInt8) -> Bool {
    guard index < bytes.count, bytes[index] == byte else { return false }
    index += 1
    return true
  }

  private mutating func skipWhitespace() {
    while index < bytes.count, [0x20, 0x09, 0x0A, 0x0D].contains(bytes[index]) {
      index += 1
    }
  }

  private func decodeString(_ range: Range<Int>) -> String? {
    let literal = String(decoding: bytes[range], as: UTF8.self)
    guard
      let values = try? JSONSerialization.jsonObject(with: Data("[\(literal)]".utf8))
        as? [String],
      values.count == 1
    else {
      return nil
    }
    return values[0]
  }

  private func isDigit(_ byte: UInt8) -> Bool {
    (0x30...0x39).contains(byte)
  }

  private func isNonzeroDigit(_ byte: UInt8) -> Bool {
    (0x31...0x39).contains(byte)
  }

  private func hexValue(_ byte: UInt8) -> UInt8? {
    if (0x30...0x39).contains(byte) { return byte - 0x30 }
    if (0x41...0x46).contains(byte) { return byte - 0x41 + 10 }
    if (0x61...0x66).contains(byte) { return byte - 0x61 + 10 }
    return nil
  }
}

File: mobile/packages/expo-mobile-web-shell/android/src/main/java/expo/modules/mobilewebshell/MobileWebExactJson.kt
package expo.modules.mobilewebshell

import org.json.JSONArray

private const val MOBILE_WEB_JSON_DEPTH_LIMIT = 32

internal fun isExactMobileWebJsonDocument(value: String): Boolean =
  MobileWebExactJsonParser(value).parse()

private class MobileWebExactJsonParser(
  private val value: String
) {
  private var index = 0

  fun parse(): Boolean {
    skipWhitespace()
    if (!parseValue(0)) return false
    skipWhitespace()
    return index == value.length
  }

  private fun parseValue(depth: Int): Boolean {
    if (depth > MOBILE_WEB_JSON_DEPTH_LIMIT || index >= value.length) return false
    return when (value[index]) {
      '{' -> parseObject(depth)
      '[' -> parseArray(depth)
      '"' -> parseStringLiteral() != null
      't' -> consume("true")
      'f' -> consume("false")
      'n' -> consume("null")
      else -> parseNumber()
    }
  }

  private fun parseObject(depth: Int): Boolean {
    if (!consume('{')) return false
    skipWhitespace()
    if (consume('}')) return true
    val keys = mutableSetOf<String>()
    while (true) {
      val literal = parseStringLiteral() ?: return false
      val key = decodeString(literal) ?: return false
      if (!keys.add(key)) return false
      skipWhitespace()
      if (!consume(':')) return false
      skipWhitespace()
      if (!parseValue(depth + 1)) return false
      skipWhitespace()
      if (consume('}')) return true
      if (!consume(',')) return false
      skipWhitespace()
    }
  }

  private fun parseArray(depth: Int): Boolean {
    if (!consume('[')) return false
    skipWhitespace()
    if (consume(']')) return true
    while (true) {
      if (!parseValue(depth + 1)) return false
      skipWhitespace()
      if (consume(']')) return true
      if (!consume(',')) return false
      skipWhitespace()
    }
  }

  private fun parseStringLiteral(): String? {
    if (!consume('"')) return null
    val start = index - 1
    while (index < value.length) {
      val character = value[index++]
      if (character == '"') return value.substring(start, index)
      if (character.code < 0x20) return null
      if (character.isHighSurrogate()) {
        if (index >= value.length || !value[index].isLowSurrogate()) return null
        index += 1
        continue
      }
      if (character.isLowSurrogate()) return null
      if (character != '\\') continue
      if (index >= value.length) return null
      val escaped = value[index++]
      if (escaped == 'u') {
        val codeUnit = parseUnicodeCodeUnit() ?: return null
        if (codeUnit in 0xD800..0xDBFF) {
          if (!consume('\\') || !consume('u')) return null
          val lowSurrogate = parseUnicodeCodeUnit() ?: return null
          if (lowSurrogate !in 0xDC00..0xDFFF) return null
        } else if (codeUnit in 0xDC00..0xDFFF) {
          return null
        }
      } else if (escaped !in "\"\\/bfnrt") {
        return null
      }
    }
    return null
  }

  private fun parseUnicodeCodeUnit(): Int? {
    if (index + 4 > value.length) return null
    var codeUnit = 0
    repeat(4) {
      val digit = value[index++].hexValue() ?: return null
      codeUnit = (codeUnit shl 4) or digit
    }
    return codeUnit
  }

  private fun parseNumber(): Boolean {
    val start = index
    consume('-')
    if (index >= value.length) return false
    if (consume('0')) {
      if (index < value.length && value[index].isJsonDigit()) return false
    } else if (!consumeDigits(firstMustBeNonzero = true)) {
      return false
    }
    if (consume('.') && !consumeDigits(firstMustBeNonzero = false)) return false
    if (index < value.length && value[index] in "eE") {
      index += 1
      if (index < value.length && value[index] in "+-") index += 1
      if (!consumeDigits(firstMustBeNonzero = false)) return false
    }
    return index > start
  }

  private fun consumeDigits(firstMustBeNonzero: Boolean): Boolean {
    if (index >= value.length) return false
    val first = value[index]
    val validFirst = if (firstMustBeNonzero) first in '1'..'9' else first.isJsonDigit()
    if (!validFirst) return false
    index += 1
    while (index < value.length && value[index].isJsonDigit()) index += 1
    return true
  }

  private fun consume(expected: String): Boolean {
    if (!value.startsWith(expected, index)) return false
    index += expected.length
    return true
  }

  private fun consume(expected: Char): Boolean {
    if (index >= value.length || value[index] != expected) return false
    index += 1
    return true
  }

  private fun skipWhitespace() {
    while (index < value.length && value[index] in " \t\n\r") index += 1
  }

  private fun decodeString(literal: String): String? = try {
    JSONArray("[$literal]").get(0) as? String
  } catch (_: Exception) {
    null
  }

  private fun Char.hexValue(): Int? = when (this) {
    in '0'..'9' -> code - '0'.code
    in 'A'..'F' -> code - 'A'.code + 10
    in 'a'..'f' -> code - 'a'.code + 10
    else -> null
  }

  private fun Char.isJsonDigit(): Boolean = this in '0'..'9'

  private fun Char.isHighSurrogate(): Boolean = code in 0xD800..0xDBFF

  private fun Char.isLowSurrogate(): Boolean = code in 0xDC00..0xDFFF
}
